hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/AutoAccountCreator.java

```diff
@@ -0,0 +1,239 @@
+
+    /**
+     * Create a transaction body for new account with the given balance and other common fields.
+     * @param balance initial balance of the account
+     * @return transaction body for new account
+     */
+    private CryptoCreateTransactionBody.Builder createAccountBase(final long balance) {
+        return CryptoCreateTransactionBody.newBuilder()
+                .initialBalance(balance)
+                .autoRenewPeriod(Duration.newBuilder().seconds(THREE_MONTHS_IN_SECONDS));
+    }
+
+    /**
+     * Create a transaction body for new account with the given alias, key, balance and maxAutoAssociations.
+     * @param alias alias of the account
+     * @param key key of the account
+     * @param balance initial balance of the account
+     * @param maxAutoAssociations maxAutoAssociations of the account
+     * @return transaction body for new account
+     */
+    private TransactionBody.Builder createAccount(
+            final Bytes alias, final Key key, final long balance, final int maxAutoAssociations) {
+        final var baseBuilder = createAccountBase(balance);
+        baseBuilder.key(key).alias(alias).memo(AUTO_MEMO);
+
+        if (maxAutoAssociations > 0) {
+            baseBuilder.maxAutomaticTokenAssociations(maxAutoAssociations);
+        }
+        return TransactionBody.newBuilder().cryptoCreateAccount(baseBuilder.build());
+    }
+
+    /**
+     * Try to recover EVM address from the given key.
+     * @param key key to recover EVM address from
+     * @param addressRecovery function to recover EVM address from the given key
+     * @return recovered EVM address if successful, otherwise null
+     */
+    @Nullable
+    private byte[] tryAddressRecovery(@Nullable final Key key, final UnaryOperator<byte[]> addressRecovery) {
+        if (key != null && key.hasEcdsaSecp256k1()) {
+            // Only compressed keys are stored at the moment
+            final var keyBytes = key.ecdsaSecp256k1OrThrow();
+            if (keyBytes.length() == ECDSA_SECP256K1_COMPRESSED_KEY_LENGTH) {
+                final var keyBytesArray = keyBytes.toByteArray();
+                final var evmAddress = addressRecovery.apply(keyBytesArray);
+                if (isRecoveredEvmAddress(Bytes.wrap(evmAddress))) {
+                    return evmAddress;
+                } else {
+                    // Not ever expected, since above checks should imply a valid input to the
+                    // LibSecp256k1 library
+                    log.warn("Unable to recover EVM address from {}", () -> hex(keyBytesArray));
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Check if the given address is of a valid EVM address length.
+     * @param address address to check
+     * @return true if the given address is a valid EVM address length, false otherwise
+     */
+    private boolean isRecoveredEvmAddress(final Bytes address) {
+        return address != null && address.length() == EVM_ADDRESS_LEN;
```

Comment:
You can use `isEvmAddressSize()` (or whatever it is, I don't have it open) here instead of `address.length() == EVM_ADDRESS_LEN`, right?

Reply:
hmm 🤔  I am not sure about that method. Where is it ?

Reply:
Ah yes, will do. Thanks